util/test/demos/d3d12/d3d12_interleaved_multi_queue.cpp

```diff
@@ -0,0 +1,459 @@
+
+      {    // Gfx work
+        // Clear RTs
+        ID3D12GraphicsCommandListPtr cmd = GetCommandBuffer();
+        Reset(cmd);
+
+        ResourceBarrier(cmd, bbTex[texIdx], D3D12_RESOURCE_STATE_PRESENT,
+                        D3D12_RESOURCE_STATE_RENDER_TARGET);
+        cmd->ClearDepthStencilView(hDsvCpuDesc, D3D12_CLEAR_FLAG_DEPTH, 0.0f, 0, 0, NULL);
+        const Vec4f clearCol(0.2f, 0.2f, 0.2f, 0.0f);
+        cmd->ClearRenderTargetView(hBackBufferRTVDescs[texIdx], &clearCol.x, 0, NULL);
+        cmd->Close();
+
+        ID3D12CommandList *rawCmdListPtr = cmd;
+        queue->ExecuteCommandLists(1, &rawCmdListPtr);
+
+        // Signal CQ0
+        TEST_ASSERT(gfxToCompute0Fence->GetCompletedValue() < sharedGfxComputeSyncCounter,
+                    "Compute0 hasn't waited for gfx signal!");
+        queue->Signal(gfxToCompute0Fence, sharedGfxComputeSyncCounter);
+
+        // Wait on CQ0
+        queue->Wait(compute0ToGfxFence, sharedGfxComputeSyncCounter);
+
+        // Signal CQ1
+        TEST_ASSERT(gfxToCompute1Fence->GetCompletedValue() < sharedGfxComputeSyncCounter,
+                    "Compute1 hasn't waited for gfx signal");
+        queue->Signal(gfxToCompute1Fence, sharedGfxComputeSyncCounter);
+
+        // Wait on CQ1
+        queue->Wait(compute1ToGfxFence, sharedGfxComputeSyncCounter);
+
+        // Draw Tris
+        cmd = GetCommandBuffer();
+        Reset(cmd);
+
+        ResourceBarrier(cmd, buffer, D3D12_RESOURCE_STATE_UNORDERED_ACCESS,
+                        D3D12_RESOURCE_STATE_PIXEL_SHADER_RESOURCE);
+
+        cmd->SetDescriptorHeaps(
+            1, &m_CBVUAVSRV.GetInterfacePtr());    // Must be done BEFORE setting root signatures
+        cmd->IASetPrimitiveTopology(D3D_PRIMITIVE_TOPOLOGY_TRIANGLELIST);
+        IASetVertexBuffer(cmd, vb, sizeof(Vec3f), 0);
+        cmd->SetPipelineState(gfx_pso);
+        cmd->SetGraphicsRootSignature(sig);
+        RSSetViewport(cmd, {0.0f, 0.0f, (float)screenWidth, (float)screenHeight, 0.0f, 1.0f});
+        RSSetScissorRect(cmd, {0, 0, screenWidth, screenHeight});
+        OMSetRenderTargets(cmd, {hBackBufferRTVDescs[texIdx]}, hDsvCpuDesc);
+        rootConstants[0] = 0;    // myBufferIdx
+        cmd->SetGraphicsRoot32BitConstants(0, 1, &rootConstants[0], 0);
+        cmd->SetGraphicsRootDescriptorTable(1, m_CBVUAVSRV->GetGPUDescriptorHandleForHeapStart());
+        cmd->DrawInstanced(3, 1, 0, 0);
+
+        rootConstants[0] = 1;    // myBufferIdx
+        cmd->SetGraphicsRoot32BitConstants(0, 1, &rootConstants[0], 0);
+        cmd->DrawInstanced(3, 1, 3, 0);
+
+        ResourceBarrier(cmd, bbTex[texIdx], D3D12_RESOURCE_STATE_RENDER_TARGET,
+                        D3D12_RESOURCE_STATE_PRESENT);
+        texIdx = 1u - texIdx;
+
+        // Transition resources back to their expected states at the start of the next frame
+        ResourceBarrier(cmd, buffer, D3D12_RESOURCE_STATE_PIXEL_SHADER_RESOURCE,
+                        D3D12_RESOURCE_STATE_UNORDERED_ACCESS);
```

Comment:
I think you want to clear the buffer here as well, since otherwise after the first loop the test doesn't actually demonstrate that the compute data was written properly before the graphics work ran - it's overwriting the same data over and over so it could produce the same output even if the replay is wrong.

Reply:
> after the first loop the test doesn't actually demonstrate that the compute data was written properly before the graphics work ran

The test now first resets the buffer to a pure blue from the gfx queue, then the compute queues read and add to the colours, adding brown and green so that the gfx pixel shading colour the triangles purple and sky blue. That's ultimately what we'd like the replay to show but, for now at least, the replay draws both tris pure blue, as expected because of the current behaviour you've described.